crates/lair_keystore_api/src/sodium_secretstream.rs

```diff
@@ -108,63 +128,75 @@ where
 pub fn new_s3_server<T, S, R>(
     send: S,
     recv: R,
-    srv_id_pub_key: sodoken::BufReadSized<32>,
-    srv_id_sec_key: sodoken::BufReadSized<32>,
+    srv_id_pub_key: Arc<[u8; 32]>,
+    srv_id_sec_key: SharedSizedLockedArray<32>,
 ) -> impl Future<Output = LairResult<(S3Sender<T>, S3Receiver<T>)>> + 'static + Send
 where
     T: 'static + serde::Serialize + for<'de> serde::Deserialize<'de> + Send,
     S: 'static + tokio::io::AsyncWrite + Send + Unpin,
     R: 'static + tokio::io::AsyncRead + Send + Unpin,
 {
     async move {
-        use sodoken::crypto_box::curve25519xchacha20poly1305 as cbox;
-        use sodoken::kx;
-
         // box these up into trait objects so we can easily refer to their types.
         let mut send: PrivRawSend = Box::new(send);
         let mut recv: PrivRawRecv = Box::new(recv);
 
         // read the sealed initiator message
-        let mut cipher: [u8; 64 + cbox::SEALBYTES] = [0; 64 + cbox::SEALBYTES];
+        let mut cipher = [0; 64 + sodoken::crypto_box::XSALSA_SEALBYTES];
         recv.read_exact(&mut cipher).await?;
-        let cipher = sodoken::BufReadSized::from(cipher);
-        let msg = <sodoken::BufWriteSized<64>>::new_no_lock();
-        cbox::seal_open(msg.clone(), cipher, srv_id_pub_key, srv_id_sec_key)
-            .await?;
-        let msg = msg.try_unwrap_sized().unwrap();
+        let mut msg = [0; 64];
+        sodoken::crypto_box::xsalsa_seal_open(
+            &mut msg,
+            &cipher,
+            &srv_id_pub_key,
+            &srv_id_sec_key.lock().lock(),
+        )?;
 
-        let oth_cbox_pub = sodoken::BufReadSized::from(&msg[..32]);
-        let oth_kx_pub = sodoken::BufReadSized::from(&msg[32..]);
+        let oth_cbox_pub: &[u8; 32] = (&msg[..32]).try_into().unwrap();
+        let oth_kx_pub: &[u8; 32] = (&msg[32..]).try_into().unwrap();
 
         // generate an ephemeral kx keypair
-        let eph_kx_pub = sodoken::BufWriteSized::new_no_lock();
-        let eph_kx_sec = sodoken::BufWriteSized::new_mem_locked()?;
-        kx::keypair(eph_kx_pub.clone(), eph_kx_sec.clone())?;
```

Comment:
Concerned by this change. We used to create a `kx::keypair`. The documentation for the `client_session_keys` below shows using a `xsalsa_keypair` instead. Is this okay? 

There is one test that seems to accidentally exercise this code, I don't see dedicated tests for it.

Reply:
Yep, kx / crypto_box / secretstream all use the same x25519 keypairs. It might be nice someday to make the api less confusing somehow.